File: Week4/Training/MVVMArchitectureComp/app/src/main/java/com/example/admin/mvvmarchitecturecomp/AppController.java
package com.example.admin.mvvmarchitecturecomp;

import android.app.Application;

import com.example.admin.mvvmarchitecturecomp.di.DaggerMainComponent;
import com.example.admin.mvvmarchitecturecomp.di.MainComponent;
import com.example.admin.mvvmarchitecturecomp.di.MainModule;

public class AppController extends Application {
    private MainComponent mainComponent;

    public MainComponent getAppComponent() {
        if (mainComponent == null)
            mainComponent = DaggerMainComponent.builder().mainModule(new MainModule(this)).build();
        return mainComponent;
    }
}

File: Week5/Homework/CodeChallengeKotlin/app/src/main/java/com/example/admin/codechallengekotlin/AppController.kt
package com.example.admin.codechallengekotlin

import android.app.Application
import com.example.admin.codechallengekotlin.di.application.ApplicationComponent
import com.example.admin.codechallengekotlin.di.application.ApplicationModule
import com.example.admin.codechallengekotlin.di.application.DaggerApplicationComponent

class AppController: Application() {
    private lateinit var applicationComponent: ApplicationComponent

    fun getComponent(): ApplicationComponent {
        return if (this::applicationComponent.isInitialized) applicationComponent else createApplicationComponent()
    }

    private fun createApplicationComponent(): ApplicationComponent {
        applicationComponent = DaggerApplicationComponent.builder().applicationModule(ApplicationModule(this)).build()
        return applicationComponent
    }
}